packages/api/portfolios/environments/createEnvironment.ts

```diff
@@ -0,0 +1,143 @@
+    const environments = await connection
+      .getRepository(Environment)
+      .createQueryBuilder("environment")
+      .select(["environment.name"])
+      .where("environment.applicationId = :applicationId", { applicationId: application.id })
+      .getMany();
+
+    // ensure the environment name is unique for the application
+    // according to business rule 3.3
+    for (const environment of environments) {
+      if (environment.name === environmentBody.name) {
+        throw createError(400, "Duplicate name");
+      }
+    }
+
+    const insertResult = await connection
+      .getCustomRepository(EnvironmentRepository)
+      .createEnvironment([{ ...environmentBody, application }]);
+
+    for (const environment of insertResult.identifiers) {
+      // the result from the insert method does not have the name property
+      // and an additional query after the insert is done to provide the correct shape
+      // to be returned to the client. This assumes that only one environment will be
+      // added at the moment but can be updated to handle more than one environment
+      const queryInsertedEnvironment = await connection.getCustomRepository(EnvironmentRepository).findOneOrFail({
+        select: [
+          "name",
+          "id",
+          "createdAt",
+          "updatedAt",
+          "archivedAt",
+          "administrators",
+          "contributors",
+          "readOnlyOperators",
+        ],
+        where: { id: environment.id },
+      });
+
+      insertedEnvironments.push(queryInsertedEnvironment);
+      console.log("Inserted Environment: " + JSON.stringify(queryInsertedEnvironment));
+    }
+  } catch (error) {
+    if (error.name === "EntityNotFoundError" || error.name === "EntityNotFoundError2") {
+      console.log("Invalid parameter entered: " + JSON.stringify(error));
+      return NO_SUCH_PORTFOLIO_OR_APPLICATION;
+    }
+    if (error.message === "Duplicate name") {
+      return DUPLICATE_ENVIRONMENT_NAME;
+    }
+    console.error("Database error: " + JSON.stringify(error));
+    return DATABASE_ERROR;
+  } finally {
+    connection.close();
+  }
+
+  return new ApiSuccessResponse<Array<IEnvironmentCreate>>(insertedEnvironments, SuccessStatusCode.CREATED);
+}
+
+// temporary work around for providing validation shape that does not fail
+// due to the 'allOf' and 'additionalProperties' conflict when composing schemas
+// see https://json-schema.org/understanding-json-schema/reference/combining.html#properties-of-schema-composition
+const environmentSchema = {
+  description: internalSchema.Environment.description,
+  type: internalSchema.Environment.type,
+  additionalProperties: internalSchema.Environment.additionalProperties,
+  required: internalSchema.Environment.required,
+  properties: {
+    ...internalSchema.Environment.properties,
+    ...internalSchema.Environment.allOf[0].properties,
+    ...internalSchema.Environment.allOf[1].properties,
+  },
+};
+
```

Comment:
Is this still a blocker?

Reply:
yeah, the same issue came up later after removing the `additionalProperties` that was in `BaseObject`.  I reverted the `BaseObject` and went with this approach. I tried the approach you suggested with the `metadata` property in the `atat-internal-api.yml`, but was not successful (not sure I did it right, though).

Reply:
I've updated the `atat-internal-api.yml` to have this structure, and this was removed. It causes duplication with `BaseObject` and `AppEnvAccess`, but ensures the schema does not have to be restructured here.